--- a/arch/arm/mach-kirkwood/Kconfig
+++ b/arch/arm/mach-kirkwood/Kconfig
@@ -205,4 +205,7 @@ source "board/alliedtelesis/SBx81LIFKW/K
 source "board/alliedtelesis/SBx81LIFXCAT/Kconfig"
 source "board/checkpoint/l-50/Kconfig"
 
+config SECOND_STAGE
+	bool "OpenWrt second stage hack"
+
 endif
--- a/include/configs/dockstar.h
+++ b/include/configs/dockstar.h
@@ -26,4 +26,6 @@
 	"console=console=ttyS0,115200\0" \
 	"bootargs_root=\0"
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_DOCKSTAR_H */
--- a/include/configs/ib62x0.h
+++ b/include/configs/ib62x0.h
@@ -29,4 +29,6 @@
 #define __io
 #endif /* CONFIG_IDE */
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_IB62x0_H */
--- a/include/configs/iconnect.h
+++ b/include/configs/iconnect.h
@@ -15,4 +15,6 @@
 	"console=console=ttyS0,115200\0"	\
 	"bootargs_root=\0"
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_ICONNECT_H */
--- /dev/null
+++ b/include/configs/openwrt-kirkwood-common.h
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2013 Luka Perkov <luka@openwrt.org>
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef __OPENWRT_KIRKWOOD_COMMON_H
+#define __OPENWRT_KIRKWOOD_COMMON_H
+
+/* Ethernet */
+#if defined(CONFIG_CMD_NET)
+#define CONFIG_SERVERIP		192.168.1.2
+#define CONFIG_IPADDR		192.168.1.1
+#endif
+
+/* second stage loader */
+#if defined(CONFIG_SECOND_STAGE)
+#undef CONFIG_ENV_IS_IN_NAND
+#undef CONFIG_ENV_SECT_SIZE
+#define CONFIG_ENV_IS_NOWHERE
+#endif
+
+/* Various */
+#define CONFIG_BZIP2
+
+/* Unnecessary */
+#undef CONFIG_BOOTM_NETBSD
+#undef CONFIG_BOOTM_PLAN9
+#undef CONFIG_BOOTM_RTEMS
+
+#endif /* __OPENWRT_KIRKWOOD_COMMON_H */
--- a/include/configs/pogo_e02.h
+++ b/include/configs/pogo_e02.h
@@ -22,4 +22,6 @@
 	"console=console=ttyS0,115200\0" \
 	"bootcmd_root=\0"
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_POGO_E02_H */
--- a/include/configs/goflexhome.h
+++ b/include/configs/goflexhome.h
@@ -35,4 +35,6 @@
 	"console=console=ttyS0,115200\0" \
 	"bootargs_root=\0"
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_GOFLEXHOME_H */
--- a/include/configs/nsa310.h
+++ b/include/configs/nsa310.h
@@ -100,4 +100,6 @@
 #define CONFIG_RTC_MV
 #endif /* CONFIG_CMD_DATE */
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_NSA310_H */
--- a/configs/dockstar_defconfig
+++ b/configs/dockstar_defconfig
@@ -54,3 +54,8 @@ CONFIG_MII=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_CMD_BOOTZ=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_LZMA=y
+CONFIG_LZO=y
--- a/configs/goflexhome_defconfig
+++ b/configs/goflexhome_defconfig
@@ -61,3 +61,8 @@ CONFIG_RTC_MV=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_CMD_BOOTZ=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_LZMA=y
+CONFIG_LZO=y
--- a/configs/ib62x0_defconfig
+++ b/configs/ib62x0_defconfig
@@ -62,5 +62,8 @@ CONFIG_SYS_NS16550_REG_SIZE=-4
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
 CONFIG_LZMA=y
+CONFIG_LZO=y
 CONFIG_BZIP2=y
--- a/configs/iconnect_defconfig
+++ b/configs/iconnect_defconfig
@@ -58,5 +58,9 @@ CONFIG_PCI_MVEBU=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_CMD_BOOTZ=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
 CONFIG_LZMA=y
+CONFIG_LZO=y
 CONFIG_BZIP2=y
--- a/configs/nsa310_defconfig
+++ b/configs/nsa310_defconfig
@@ -43,6 +43,9 @@ CONFIG_CMD_UBI=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_CMD_BOOTZ=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
 CONFIG_LZMA=y
 CONFIG_LZO=y
 CONFIG_SYS_LONGHELP=y
--- a/configs/pogo_e02_defconfig
+++ b/configs/pogo_e02_defconfig
@@ -54,3 +54,8 @@ CONFIG_MII=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_CMD_BOOTZ=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_LZMA=y
+CONFIG_LZO=y
--- a/configs/nsa310s_defconfig
+++ b/configs/nsa310s_defconfig
@@ -58,5 +58,8 @@ CONFIG_DEBUG_UART_SHIFT=2
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_UBIFS_SILENCE_MSG=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
 CONFIG_LZMA=y
+CONFIG_LZO=y
 CONFIG_BZIP2=y
--- a/include/configs/nsa310s.h
+++ b/include/configs/nsa310s.h
@@ -31,4 +31,6 @@
 	"bootargs_root=\0" \
 	LOAD_ADDRESS_ENV_SETTINGS
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_NSA310S_H */
--- a/configs/sheevaplug_defconfig
+++ b/configs/sheevaplug_defconfig
@@ -69,5 +69,8 @@ CONFIG_RTC_MV=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_STORAGE=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
 CONFIG_LZMA=y
+CONFIG_LZO=y
 CONFIG_BZIP2=y
